### PR TITLE
feat(megatron): add nccl_comm_warmup to avoid iteration-1 NCCL cudaMalloc OOM (#6387)

### PR DESCRIPTION
**Pull request by yuchenwang3**
## What

Adds an opt-in `nccl_comm_warmup` flag that eagerly creates the NCCL communicators before the training loop, to avoid an iteration-1 OOM.

## Why

NCCL communicators are created lazily on first use. For the per-step collectives that ms-swift/Megatron issue inside the training step (e.g. the dp/cp loss all-reduce and grad-sync coalescing), that first use happens at the **iteration-1 memory peak**. At that point NCCL's internal `cudaMalloc` for the communicator buffers can fail with:

```
Failed to CUDA calloc async N bytes
```

so training dies on the very first step on large models / tight-memory configs. This is the error reported in #6387 (Qwen3-Next-80B-A3B, 8×H20), which was only worked around by changing the parallel layout (`--expert_model_parallel_size 8`) rather than fixed in code. We hit the same failure on Qwen3.5-35B-A3B.

## Change

When `--nccl_comm_warmup true`, right before `on_train_begin` (in `setup_model_training`), fire one numerically-inert 1-element `all_reduce` on each parallel group (dp, dp+cp, cp, tp, pp, model, embedding, position-embedding). This forces the communicators to be allocated while memory is still free; groups that don't exist for the current parallel config are skipped. Default is `False`, so behavior is unchanged unless explicitly enabled.

## Notes

- Opt-in and default-off; zero effect on existing runs.
- The dummy all-reduce is a no-op numerically (a 1-element zero tensor).
- Verified the two files compile; could not run a multi-GPU job locally — relying on CI / users who hit #6387.

### CODE REVIEW

**Comment by yuchenwang3** [comment]
@Jintao-Huang gentle ping 🙏 Small self-contained `nccl_comm_warmup` flag to avoid the first-iteration NCCL hang, default off. Would appreciate a quick look when you have time.

**Review by Jintao-Huang — APPROVED** [review]

**Comment by Jintao-Huang** [comment]
thanks!
